test/x509_check_cert_pkey_test.c

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2017 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdio.h>
+#include <string.h>
+
+#include <openssl/pem.h>
+#include <openssl/x509.h>
+#include "testutil.h"
+
+/*
+ * c: path of a cert in PEM format
+ * k: path of a key in PEM format
+ * t: API type, "cert" for X509_ and "req" for X509_REQ_ APIs.
+ * e: expected, "ok" for success, "failed" for what should fail.
+ */
+static int test_x509_check_cert_pkey(const char *c, const char *k,
+    const char *t, const char *e)
+{
+    BIO *bio = NULL;
+    X509 *x509 = NULL;
+    X509_REQ *x509_req = NULL;
+    EVP_PKEY *pkey = NULL;
+    int ret = 0, type = 0, expected = 0, result = 0;
+
+    /*
+     * we check them first thus if fails we don't need to do
+     * those PEM parsing operations.
+     */
+    if (strcmp(t, "cert") == 0) {
+        type = 1;
+    } else if (strcmp(t, "req") == 0) {
+        type = 2;
+    } else {
+        TEST_error("invalid 'type'");
+        goto failed;
+    }
+
+    if (strcmp(e, "ok") == 0) {
+        expected = 1;
+    } else if (strcmp(e, "failed") == 0) {
+        expected = 2;
+    } else {
+        TEST_error("invalid 'expected'");
+        goto failed;
+    }
+
+    /* process private key */
+    bio = BIO_new_file(k, "r");
+    if (bio == NULL) {
```

Comment:
A style more consistent with the rest of the tests would be:

    if (!TEST_ptr(bio = BIO_new_file(k, "r")))
        goto failed